native_client/deepspeech.cc

```diff
@@ -21,347 +23,670 @@
     decoder.Decode(seq_len, inputs, &decoder_outputs, &scores).ok();
   }
 
-  // Output is an array of shape (1, n_results, result_length).
-  // In this case, n_results is also equal to 1.
-  size_t output_length = decoder_outputs[0][0].size() + 1;
+  // Output is an array of shape (batch_size, top_paths, result_length).
 
-  size_t decoded_length = 1; // add 1 for the \0
-  for (int i = 0; i < output_length - 1; i++) {
-    int64 character = decoder_outputs[0][0][i];
-    const std::string& str = mPriv->alphabet->StringFromLabel(character);
-    decoded_length += str.size();
+  std::stringstream output;
+  for (int64 character : decoder_outputs[0][0]) {
+    output << alphabet->StringFromLabel(character);
   }
 
-  char* output = (char*)malloc(sizeof(char) * decoded_length);
-  char* pen = output;
-  for (int i = 0; i < output_length - 1; i++) {
-    int64 character = decoder_outputs[0][0][i];
-    const std::string& str = mPriv->alphabet->StringFromLabel(character);
-    strncpy(pen, str.c_str(), str.size());
-    pen += str.size();
+  return strdup(output.str().c_str());
+}
+
+int
+DS_CreateModel(char* aModelPath,
+               int aNCep,
+               int aNContext,
+               char* aAlphabetConfigPath,
+               int aBeamWidth,
+               ModelState** retval)
+{
+  ModelState* model = new ModelState();
+  model->mmap_env   = new MemmappedEnv(Env::Default());
+  model->ncep       = aNCep;
+  model->ncontext   = aNContext;
+  model->alphabet   = new Alphabet(aAlphabetConfigPath);
+  model->beam_width = aBeamWidth;
+  model->run_aot    = false;
+
+  *retval = nullptr;
+
+  DS_PrintVersions();
+
+  if (!aModelPath || strlen(aModelPath) < 1) {
+    std::cerr << "No model specified, will rely on built-in model." << std::endl;
+    model->run_aot = true;
+    return 0;
+  }
+
+  Status status;
+  SessionOptions options;
+
+  bool is_mmap = std::string(aModelPath).find(".pbmm") != std::string::npos;
+  if (!is_mmap) {
+    std::cerr << "Warning: reading entire model file into memory. Transform model file into an mmapped graph to reduce heap usage." << std::endl;
+  } else {
+    status = model->mmap_env->InitializeFromFile(aModelPath);
+    if (!status.ok()) {
+      std::cerr << status << std::endl;
+      return status.code();
+    }
+
+    options.config.mutable_graph_options()
+      ->mutable_optimizer_options()
+      ->set_opt_level(::OptimizerOptions::L0);
+    options.env = model->mmap_env;
+  }
+
+  status = NewSession(options, &model->session);
+  if (!status.ok()) {
+    std::cerr << status << std::endl;
+    return status.code();
+  }
+
+  if (is_mmap) {
+    status = ReadBinaryProto(model->mmap_env,
+                             MemmappedFileSystem::kMemmappedPackageDefaultGraphDef,
+                             &model->graph_def);
+  } else {
+    status = ReadBinaryProto(Env::Default(), aModelPath, &model->graph_def);
+  }
+  if (!status.ok()) {
+    std::cerr << status << std::endl;
+    delete model;
```

Comment:
Super bug prone. Should use smart pointers[[1](https://stackoverflow.com/a/395158/1297739)].